mlc_llm/core.py

```diff
@@ -311,22 +320,27 @@ def mod_transform_before_build(
     if "num_attention_heads" in config and "hidden_size" in config:
         mod = fuse_split_rotary_embedding(mod, config["num_attention_heads"], config["hidden_size"])
 
-    if args.target_kind == "cuda" and tvm.get_global_func("relax.ext.cutlass", True):
-        # CUTLASS offloading
+    if args.target_kind == "cuda":
         patterns = []
 
-        if not args.no_cutlass_attn:
+        has_cutlass = tvm.get_global_func("relax.ext.cutlass", True)
+        has_cublas = tvm.get_global_func("relax.ext.cublas", True)
+
+        if has_cutlass and not args.no_cutlass_attn:
             mod["prefill"] = rewrite_attention(mod["prefill"])
             mod["decode"] = rewrite_attention(mod["decode"])
             patterns += get_patterns_with_prefix("cutlass.attention")
 
-        if not args.no_cutlass_norm:
+        if has_cutlass and not args.no_cutlass_norm:
             patterns += get_patterns_with_prefix("cutlass.layer_norm")
             patterns += get_patterns_with_prefix("cutlass.rms_norm")
 
-        if use_ft_quant:
+        if has_cutlass and use_ft_quant:
             patterns += get_patterns_with_prefix("cutlass.decode_matmul")
 
+        if has_cublas and args.use_cublas:
```

Comment:
Enable only when q0f16?

Reply:
+1

Reply:
Can we enable it by default for q0f16?

Reply:
I updated the code to enable cublas offload in q0f16 and q0f32, and changed the flag to `no_cublas`.